src/prng/hc128.rs

```diff
@@ -0,0 +1,543 @@
+            (y << 32) | x
+        }
+    }
+
+    // As an optimization we try to write directly into the output buffer.
+    // This is only enabled for platforms where unaligned writes are known to
+    // be safe and fast.
+    // This improves performance by about 12%.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        let mut filled = 0;
+
+        // Continue filling from the current set of results
+        if self.index < 16 {
+            let (consumed_u32, filled_u8) =
+                impls::fill_via_u32_chunks(&mut self.results[self.index..],
+                                           dest);
+
+            self.index += consumed_u32;
+            filled += filled_u8;
+        }
+
+        let len_remainder = (dest.len() - filled) % 16;
+        let len_direct = dest.len() - len_remainder;
+
+        while filled < len_direct {
+            let dest_u32: &mut [u32] = unsafe {
+                slice::from_raw_parts_mut(
+                    dest[filled..].as_mut_ptr() as *mut u8 as *mut u32,
+                                      16)
+            };
+            self.state.update(dest_u32);
+            filled += 16 * 4;
+        }
+        self.index = 16;
+
+        if len_remainder > 0 {
+            self.state.update(&mut self.results);
+
+            let (consumed_u32, _) =
+                impls::fill_via_u32_chunks(&mut self.results,
+                                           &mut dest[filled..]);
+
+            self.index = consumed_u32;
+        }
+    }
+
+    #[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        let mut read_len = 0;
+        while read_len < dest.len() {
+            if self.index >= 16 {
+                self.state.update(&mut self.results);
+                self.index = 0;
+            }
+
+            let (consumed_u32, filled_u8) =
+                impls::fill_via_u32_chunks(&mut self.results[self.index..],
+                                           &mut dest[read_len..]);
+
+            self.index += consumed_u32;
+            read_len += filled_u8;
+        }
+    }
+}
+
+impl Rand for Hc128Rng {
+    fn rand<R: Rng>(other: &mut R) -> Hc128Rng {
+        let mut seed = [0u32; 8];
+        unsafe {
+            let ptr = seed.as_mut_ptr() as *mut u8;
+            let slice = slice::from_raw_parts_mut(ptr, 8 * 4);
+            other.fill_bytes(slice);
+        }
+        Hc128Rng::init(&seed)
+    }
+}
+
+impl<'a> SeedableRng<&'a [u32]> for Hc128Rng {
+    fn reseed(&mut self, seed: &'a [u32]) {
+        *self = Self::from_seed(seed);
+    }
+    /// Create an HC-128 random number generator with a seed. The seed has to be
+    /// 256 bits in length, matching the 128 bit `key` followed by 128 bit `iv`
+    /// when HC-128 where to be used as a stream cipher.
+    fn from_seed(seed: &'a [u32]) -> Hc128Rng {
+        assert!(seed.len() == 8);
```

Comment:
Why use a slice when the length must be fixed? Why not use an array ref? I know the current convention is to use a slice, but we don't need to respect that.

Reply:
Oh, we discussed that! Fixed it in some branch, but forget to update here. https://github.com/dhardy/rand/pull/77/files#diff-9c520b54b9848356a0cf51d6f963749cL87

Reply:
Should I change the seed type to `[u8, 32]`, or make it `[u32; 8]`? With an `u8` seed I would also need to include the `read_u32_into` function.